pcdscalc/pmps.py

```diff
@@ -0,0 +1,256 @@
+        use the soft-xray bitmask.
+
+    Returns
+    -------
+    boundaries: list of floats
+    """
+    if line.lower()[0] in ['l', 'h']:
+        return LFE
+    if line.lower()[0] in ['k', 's']:
+        return KFE
+    raise ValueError(f'{line} is neither lfe or kfe!.')
+
+
+def get_bitmask(lower, upper, allow, line, bounds=None):
+    """
+    Given a range of eV values, calculate the appropriate pmps bitmask.
+
+    This saves you the effort of checking up on the eV ranges and
+    remembering how the bitmask is assembled.
+
+    If the lower or upper fall within a range (not on the border), then that
+    range is always considered unsafe.
+
+    The rules for the bitmasks are:
+        - The nth bit of the bitmask represents a range from the n-1th value
+          to the nth value
+        - If the bit is 1, the range is allowed and is safe.
+        - If the bit is 0, the range is not allowed and is not safe.
+        - eVs above the highest boundary are never safe
+        - negative eVs are never safe
+
+    Some examples:
+        - bitmask 00 -> no beam range is safe
+        - bitmask 01 -> only the lowest boundary and below are safe
+        - bitmask 11 -> only the two lowest boundaries and below are safe
+        - bitmask 10 -> only the space between the lowest and second-lowest
+                        boundaries are safe.
+        - bitmask all 1s -> every eV is OK except for above the highest
+                            boundary and except for negative eV
+
+    Parameters
+    ----------
+    lower: number
+        The value in eV for the lower bound of the range.
+
+    upper: number
+        The value in eV for the upper bound of the range.
+
+    allow: bool
+        True if we want a bitmask that only includes this range,
+        False if we want a bitmask that only excludes this range.
+
+    line: str
+        String representation of which line's bitmask to use.
+        If the string begins with "l" or "h" (lfe, hxr), we'll
+        use the hard-xray bitmask.
+        If the string begins with "k" or "s" (kfe, sxr), we'll
+        use the soft-xray bitmask.
+
+    bounds: list of numbers, optional
+        Custom boundaries to use instead of the default soft-xray
+        or hard-xray lines. Useful for testing.
+
+    Returns
+    -------
+    bitmask: int
+    """
+    # Be lenient on input args
+    if upper < lower:
+        lower, upper = upper, lower
+
+    if allow:
+        bounds = bounds or select_bitmask_boundaries(line)
+        bitmask = 0
+
+        prev = 0
+        for bit, ev in enumerate(bounds):
+            if lower <= prev and upper >= ev:
+                bitmask += 2**bit
+            prev = ev
+        return bitmask
+
+    # An exclusion range is just two inclusion ranges
+    else:
+        lower_range = get_bitmask(0, lower, True, line, bounds=bounds)
+        upper_range = get_bitmask(upper, math.inf, True, line, bounds=bounds)
+        return lower_range | upper_range
```

Comment:
Neat usage here! 

One thought:
* `get_bitmask(allow=True)` is inclusive at its bounds `[lower, upper]`.
* Noting that I haven't done the math here so I could easily be off-base -- would this let through `lower` and `upper`? And if so, is that intentional? The docstring implies they should be excluded.

Reply:
`get_bitmask(allow=True)` is intended to exclude the endpoints unless they match an eV boundary exactly.

I'm going to talk through the test cases to make sure they all make sense and I didn't miss an edge case (and because this is confusing):

```
     (0, 15.5, True,     0b0000_0000_0000_0000_0111_1111_1111_1111),
     (15.5, 100, True,   0b1111_1111_1111_1111_0000_0000_0000_0000),
     (14.5, 21.5, True,  0b0000_0000_0001_1111_1000_0000_0000_0000),
     (14.5, 21.5, False, 0b1111_1111_1100_0000_0011_1111_1111_1111),
     (15, 20, True,      0b0000_0000_0000_1111_1000_0000_0000_0000)
```
```
boundaries = [1, 2, 3, 4 ... 32]
```

1. Allow between 0 and 15, exclude the 15.5 point because we can't allow points like 15.6 and they share a range. Therefore, enable bits 1 through 15 (bit 1 allows 0 to 1, bit 15 allows 14 to 15)
2. Allow between 16 and 32, exclude the 15.5 point because we can't allow points like 15.4 and they share a range. Cut off at 32 for the top of the bitmask range. Therefore, enable bits 17 through 32 (bit 17 allows 16 to 17)
3. Allow between 15 and 21, exclude the exact points because we can't allow any points outside the range. Enable bits 16 to 21 (bit 16 allows 15 to 16, bit 21 allows 20 to 21)
4. Allow between 0 and 14 and between 22 and 32. Turn off bits 15 to 22 (bit 15 allows 14 to 15, bit 22 allows 21 to 22). Exclude the boundary bits for the same reasonings as above.
5. Allow exactly between 15 and 20. Turn on bits 16 to 20 (bit 16 allows 15 to 16, bit 20 allows 19 to 20)

Now, perhaps I'm missing the case of an exact range with False? Perhaps:
```
     (15, 20, False,     0b1111_1111_1111_0000_0111_1111_1111_1111)
```
Which turns off bits 16 to 20 (bit 16 allows 15 to 16, bit 20 allows 19 to 20)
Checking the algorithm:
```
In [8]: bin(get_bitmask(15, 20, False, 'tst', list(range(1,33))))
Out[8]: '0b11111111111100000111111111111111'
# 0b1111_1111_1111_0000_0111_1111_1111_1111
```
Ok, I think I'm satisfied with that one.

Is there some other case I'm missing?

Reply:
The only edge cases seem to be where they are on the exact boundary, which you've described.

Reading through your logic, it makes enough sense to me to be confident you've covered all the bases - thanks for that. 👍

Reply:
:+1: I'm going to add most of that text as a block comment in the test suite